### ISSUE TITLE
Add AnnotationConfigServletWebServerApplicationContext construtor that takes a beanFactory  parameter

### ISSUE THREAD

**Issue by xnat9** [spring-projects/spring-boot#8547]
`AnnotationConfigEmbeddedWebApplicationContext` can not use a `CustomBeanFactory`???

`CustomBeanFactory extends DefaultListableBeanFactory`

**Comment by xnat9** [comment]
custm

**Comment by wilkinsona** [comment]
That's correct. Why do you want to use a custom bean factory?

**Comment by xnat9** [comment]
override some method in DefaultListableBeanFactory. toString method in DefaultListableBeanFactory will print too many info, i want to overidde toString method in DefaultListableBeanFactory.
why can not custom BeanFactory?

**Comment by wilkinsona** [comment]
Sorry, but customising the value returned from `toString` isn't sufficient justification to add the complexity of allowing the bean factory used by AnnotationConfigEmbeddedWebApplicationContext to be customised. Spring Framework's AnnotationConfigWebApplicationContext doesn't allow it either.

**Comment by binnchx** [comment]
@wilkinsona , I also need a custom BeanFactory, because I want to speed up the application startup time,  so  I  override the default invokeInitMethods of the DefaultListableBeanFactory  and use threadpool to parallel initializing beans during startup.
https://jira.spring.io/browse/SPR-8767

**Comment by wilkinsona** [comment]
Sorry, but that doesn’t feel like sufficient reason either. What you are doing is an edge case. Awaiting official support in Spring Framework is preferable to adding complexity to Spring Boot that few will use.

**Comment by binnchx** [comment]
@wilkinsona , I can use custom beanfactory in spring application, but why I can't in spring boot application?

**Comment by philwebb** [comment]
We should probably add additional constructors to `AnnotationConfigServletWebServerApplicationContext` and `ServletWebServerApplicationContext` to align with `GenericWebApplicationContext`.

**Comment by philwebb** [comment]
Also the reactive contexts as well

**Comment by wilkinsona** [comment]
For the record, I was wrong before when I said that `AnnotationConfigWebApplicationContext` doesn't support bean factory customisation. It does support it via subclassing and overriding `createBeanFactory()`.

**Comment by philwebb** [comment]
Good spot. So `AnnotationConfigReactiveWebApplicationContext` should be OK and we can add inherited constructors for the rest.

**Comment by binnchx** [comment]
hello, @philwebb , in spring boot 1.4.3.RELEASE, `AnnotationConfigReactiveWebApplicationContext ` doesn't exist?

**Comment by wilkinsona** [comment]
@binnchx That's right. It's new in 2.0 in support of Boot's Spring WebFlux integration.

**Comment by snicoll** [comment]
@philwebb for the record, I don't understand why we have to do this. The fact the framework as advanced hook points does not mean we have to add them. I very much prefer to hear why someone has to override the `BeanFactory` rather than letting them do that. Perhaps there is an improvement to be made that would be applicable to more users?

**Comment by philwebb** [comment]
> for the record, I don't understand why we have to do this. The fact the framework as advanced hook points does not mean we have to add them

I guess the question for me is why wouldn't we do this? It's a relatively small change on our side and it brings consistency with the framework. Users still have to jump through some hoops if the want to plug-in a custom `BeanFactory`, but at least they now can. I don't think the addition constructor will be taken as a sign that we expect people to use it often.

I think the requests listed above are certainly edge-cases, but I'd still like people to have the option to experiment if they'd like to. I might feel differently if were a lot if work to support this, but it really was just a 10min job.

> I very much prefer to hear why someone has to override the BeanFactory rather than letting them do that. Perhaps there is an improvement to be made that would be applicable to more users?

I imagine experimentation will be the biggest one. Parallel bean initialization mentioned above sounds interesting. I've also often messed with the `BeanFactory` to add timing hooks when looking at performance issues.

**tl;dr** : It was a small change and I can't see many downsides.
